### PR TITLE
Widen type promotion for decimals with larger scale in Parquet Read [databricks]

### PR DESCRIPTION
**Pull request by nartal1**
This PR contributes to https://github.com/NVIDIA/spark-rapids/issues/11433 and contributes to https://github.com/NVIDIA/spark-rapids/issues/11512

This PR supports additional type promotion to decimals with larger precision and scale.
As long as the precision increases by at least as much as the scale, the decimal values can be promoted without loss of precision. 
A similar change is added in Apache Spark-4.0 version - https://github.com/apache/spark/pull/44513
Currently, the code throws an Exception if the scale of read schema is not as same as the schema that was written for all versions previous to Spark-4.0 on CPU. This fix is available for all versions in spark-rapids.

We have removed separate checks for the decimal  if they can be read as int, long and byte_array and consolidated into one function `canReadAsDecimal`. Added integration test to verify that the conditions of the type promotions are met.

### CODE REVIEW

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 1]
```diff
@@ -787,19 +787,19 @@ private case class GpuParquetFileFilterHandler(
...
+   * @param rootFileType  file type of each root column
+   * @param rootReadType  read type of each root column
    */
   private def checkSchemaCompat(fileType: Type,
-                                readType: DataType,
-                                errorCallback: (Type, DataType) => Unit,
-                                isCaseSensitive: Boolean,
-                                useFieldId: Boolean,
-                                rootFileType: Option[Type] = None,
-                                rootReadType: Option[DataType] = None): Unit = {
```
Comment (mythrocks) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:802]: We badly need a code formatter.  Or a better one.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 2]
```diff
@@ -1031,33 +1027,25 @@ private case class GpuParquetFileFilterHandler(
...
-    DecimalType.isByteArrayDecimalType(dt) && isDecimalTypeMatched(pt.getDecimalMetadata, dt)
+  private def canReadAsDecimal(pt: PrimitiveType, dt: DataType): Boolean = {
+    (DecimalType.is32BitDecimalType(dt)
+      || DecimalType.is64BitDecimalType(dt)
+      || DecimalType.isByteArrayDecimalType(dt)) && isDecimalTypeMatched(pt.getDecimalMetadata, dt)
   }
 
   // TODO: After we deprecate Spark 3.1, fetch decimal meta with DecimalLogicalTypeAnnotation
   @scala.annotation.nowarn("msg=class DecimalMetadata in package schema is deprecated")
   private def isDecimalTypeMatched(metadata: DecimalMetadata,
```
Comment (mythrocks) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1040]: I would have named this `isDecimalTypeCompatible`, but it looks like Apache Spark names it this way as well.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 3]
```diff
@@ -944,30 +945,30 @@ private case class GpuParquetFileFilterHandler(
...
+        if (dt == DataTypes.ByteType || dt == DataTypes.ShortType || dt == DataTypes.DateType) {
+          return
+        }
 
       case PrimitiveTypeName.INT64 =>
         if (dt == DataTypes.LongType || GpuTypeShims.isSupportedDayTimeType(dt) ||
-            // Day-time interval type is stored as int64 in parquet
-            canReadAsLongDecimal(pt, dt)) {
+          // Day-time interval type is stored as int64 in parquet
+          canReadAsDecimal(pt, dt)) {
```
Comment (mythrocks) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:971]: 👍

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 4]
```diff
@@ -1031,33 +1027,25 @@ private case class GpuParquetFileFilterHandler(
...
-                                   sparkType: DataType): Boolean = {
+      sparkType: DataType): Boolean = {
     if (metadata == null) {
       false
     } else {
       val dt = sparkType.asInstanceOf[DecimalType]
-      metadata.getPrecision <= dt.precision && metadata.getScale == dt.scale
+      val scaleIncrease = dt.scale - metadata.getScale
+      val precisionIncrease = dt.precision - metadata.getPrecision
+      scaleIncrease >= 0 && precisionIncrease >= scaleIncrease
```
Comment (mythrocks) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1048]: This seems to be the crux of it. 👍

**Review by mythrocks — APPROVED** [review]
I'm 👍 on this change.

A fair bit of this change involves fixing whitespaces.  We should see about a code formatter, after this change is merged.

File: integration_tests/src/main/python/parquet_test.py [hunk 5]
```diff
@@ -1481,6 +1474,71 @@ def test_parquet_check_schema_compatibility_nested_types(spark_tmp_path):
...
+    # Write test data with CPU
+    with_cpu_session(
+        lambda spark: unary_op_df(spark, from_decimal_gen)
+        .coalesce(1)
+        .write.parquet(data_path)
+    )
+
+    # Create target schema for reading
+    read_schema = StructType([
+        StructField("a", DecimalType(to_decimal_gen.precision, to_decimal_gen.scale))
```
Comment (revans2) [inline integration_tests/src/main/python/parquet_test.py:1515]: nit: `to_decimal_gen.data_type` should return for you the same as the DecimalType shown there.

File: integration_tests/src/main/python/parquet_test.py [hunk 6]
```diff
@@ -1481,6 +1474,71 @@ def test_parquet_check_schema_compatibility_nested_types(spark_tmp_path):
...
+    # Determine if we expect an error based on precision and scale changes
+    expect_error = (
+            to_decimal_gen.scale < from_decimal_gen.scale or
+            (to_decimal_gen.precision - to_decimal_gen.scale) <
+            (from_decimal_gen.precision - from_decimal_gen.scale)
+    )
+
+    spark_conf = {}
+    if is_before_spark_400():
+        spark_conf['spark.sql.parquet.enableVectorizedReader'] = 'false'
```
Comment (revans2) [inline integration_tests/src/main/python/parquet_test.py:1527]: Can we have a comment here about why this is set?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 7]
```diff
@@ -944,30 +945,30 @@ private case class GpuParquetFileFilterHandler(
    */
   @scala.annotation.nowarn("msg=method getDecimalMetadata in class PrimitiveType is deprecated")
   private def checkPrimitiveCompat(pt: PrimitiveType,
-                                   dt: DataType,
-                                   errorCallback: () => Unit): Unit = {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:948]: nit: most of the changes to this file are formatting and have nothing to do with the changes being made. Can we please revert these so there is less of a chance of code conflicts?

**Review by revans2 — APPROVED** [review]
Just some nits.

**Comment by nartal1** [comment]
@revans2 @mythrocks  I have addressed the review comments. Reverted the formatting changes. PTAL.

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Review by mythrocks — APPROVED** [review]

**Pull request merged**